File: app/src/main/java/com/netspeed/monitor/BootReceiver.java
package com.netspeed.monitor;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Build;
import android.util.Log;

/**
 * Listens for device boot events and restarts the speed monitoring service
 * if the user enabled "Start on Boot" in settings.
 */
public class BootReceiver extends BroadcastReceiver {

    // Tag for logging boot receiver events
    private static final String TAG = "BootReceiver";

    @Override
    public void onReceive(Context context, Intent intent) {
        // Only respond to boot-related broadcasts
        String action = intent.getAction();
        if (!Intent.ACTION_BOOT_COMPLETED.equals(action) &&
                !"android.intent.action.QUICKBOOT_POWERON".equals(action)) {
            return;
        }

        Log.d(TAG, "Boot broadcast received: " + action);

        // Read the start-on-boot preference from SharedPreferences
        SharedPreferences prefs = context.getSharedPreferences("settings", Context.MODE_PRIVATE);
        boolean startOnBoot = prefs.getBoolean("start_on_boot", false);

        if (startOnBoot) {
            // Build intent to start the foreground speed monitoring service
            Intent serviceIntent = new Intent(context, SpeedMonitorService.class);
            serviceIntent.setAction(SpeedMonitorService.ACTION_START);

            try {
                // startForegroundService is required on API 26+ (our minSdk)
                context.startForegroundService(serviceIntent);
                Log.d(TAG, "Foreground service start requested successfully");
            } catch (Exception e) {
                // On Android 12+ a ForegroundServiceStartNotAllowedException may be thrown
                // if the system denies the background start despite the BOOT_COMPLETED exemption
                Log.e(TAG, "Failed to start foreground service on boot: " + e.getMessage(), e);
            }
        } else {
            Log.d(TAG, "Start on boot is disabled, skipping service start");
        }
    }
}

File: app/src/main/java/com/netspeed/monitor/receiver/BootReceiver.kt
package com.netspeed.monitor.receiver

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.os.Build
import com.netspeed.monitor.service.SpeedMonitorService
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import androidx.datastore.preferences.core.booleanPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import kotlinx.coroutines.flow.first

// Extension to get the settings DataStore (same name as PreferencesRepositoryImpl)
private val Context.dataStore by preferencesDataStore(name = "settings")

// BroadcastReceiver that listens for device boot events and restarts the monitoring service
class BootReceiver : BroadcastReceiver() {

    // Called when BOOT_COMPLETED or QUICKBOOT_POWERON intent is received
    override fun onReceive(context: Context, intent: Intent) {
        // Verify this is a boot-related broadcast before doing anything
        if (intent.action != Intent.ACTION_BOOT_COMPLETED &&
            intent.action != "android.intent.action.QUICKBOOT_POWERON"
        ) {
            return
        }

        // Use goAsync() to extend the time window for async DataStore read
        val pendingResult = goAsync()

        // Launch a coroutine on IO dispatcher to check the user's start-on-boot preference
        CoroutineScope(Dispatchers.IO).launch {
            try {
                // Read the start-on-boot preference from DataStore
                val prefs = context.dataStore.data.first()
                val startOnBoot = prefs[booleanPreferencesKey("start_on_boot")] ?: false

                // Only start the service if the user has enabled start-on-boot
                if (startOnBoot) {
                    // Build the intent to start the foreground monitoring service
                    val serviceIntent = Intent(context, SpeedMonitorService::class.java).apply {
                        action = SpeedMonitorService.ACTION_START
                    }
                    // Use startForegroundService for Android 8.0+ to satisfy foreground requirements
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                        context.startForegroundService(serviceIntent)
                    } else {
                        context.startService(serviceIntent)
                    }
                }
            } finally {
                // Signal Android that the BroadcastReceiver has finished async processing
                pendingResult.finish()
            }
        }
    }
}
